build: avoid building test bins with duplicate TAGS, GOFLAGS, LINKFLAGS

knz opened the pull request:
Fixes #21086.

Recursively calling `$(MAKE) TAGS='$(TAGS)'` (ditto GOFLAGS,
LINKFLAGS) causes the various strings to be appended twice to these
variables (once in the caller make, one in the callee make) and then
passed twice to the actual go commands. This causes some troubles on
e.g. freebsd (see issue #21086).

This patch alleviates the issue by removing the explicit variable
overrides from the recursive make invocation.

I have checked manually that if a user runs `make testlogic TAGS=foo`
then the custom TAGS is properly propagated to the recursive make.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/21092)
<!-- Reviewable:end -->

benesch commented:
Ugh, sorry, this is my bad for trying to reduce the duplication in our Makefile. I’d prefer to see if `make testbuild PKG='./pkg/$(*D)'` would do the trick rather than reintroducing the duplication, but feel free to merge this as-is. I’m not sure whether my proposed solution would have the desired effect.

If you do merge as is, @knz, can you file an issue to kill the testbuild target? Every testbuild invocation can be trivially rewritten in terms of the `pkg/%.test` target. That’ll solve the duplication problem and avoid recursive make, so I think it’s the right long term solution.

knz commented:
Nikhil you're right just calling `make testbuild PKG=./pkg/$(*D)` works properly. I changed the patch to do just that. PTAL.

knz committed:
build: avoid building test bins with duplicate TAGS, GOFLAGS, LINKFLAGS

Recursively calling `$(MAKE) TAGS='$(TAGS)'` (ditto GOFLAGS,
LINKFLAGS) causes the various strings to be appended twice to these
variables (once in the caller make, one in the callee make) and then
passed twice to the actual go commands. This causes some troubles on
e.g. freebsd (see issue #21086).

This patch alleviates the issue by removing the explicit variable
overrides from the recursive make invocation.

I have checked manually that if a user runs `make testlogic TAGS=foo`
then the custom TAGS is properly propagated to the recursive make.

Release note: None

benesch commented:
🙌🏽 LGTM!

On Thu, Dec 28, 2017 at 12:01 PM kena <notifications@github.com> wrote:

> Nikhil you're right just calling make testbuild PKG=./pkg/$(*D) works
> properly. I changed the patch to do just that. PTAL.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/21092#issuecomment-354329123>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA15IJ19p91Jztp9gWpalDSFWuANAx9cks5tE9eJgaJpZM4ROfhW>
> .
>
